src/kernels/jupyter/launcher/serverUriStorage.ts

```diff
@@ -28,16 +31,32 @@ export class JupyterServerUriStorage implements IJupyterServerUriStorage {
     public get onDidRemoveUris() {
         return this._onDidRemoveUris.event;
     }
+    public get currentUri(): string | undefined {
+        return this._currentUri;
```

Comment:
We could have issues due to race conditions.
Basically `getUri` is async, hence `currentUri` could be empty when it shouldn't be and could give the wrong impression when its used.

Reply:
Isn't this a duplicate of  the `getUri` method, without the async bit.

Reply:
Yes. The sync version. getUri just uses this value once it's set anyway.

Reply:
I think we can remove this, we don't need a suync version, we can make this async &
just make other parts of the code async when accessing this.
From what I could tell we use this in adding controllers, and we need to do is turn the `IControllerRegistration.add` method into an async method.
From what I could tell that should be totally fine, as almost every method that calls the `add` method is already async.

Reply:
From what I can tell we only need `currentUri` for filtering:
* We could use the `getUri` async method & turn all others into async
* Or just use the `serverId` (we can save server Id in memto as that's safe & hashed, and its part of the connection metadata, & since thats all synchronous, we can use that via `currentServerId` property & then use that for filtering)

Reply:
Wait what's serverId? Is that the hash of the connection information.

Reply:
Hmm. ServerId is a hash of the current uri. It's not secure but it's only in memory. If we wrote it to a memento then there's a risk of leaking somebody's token. I guess there's a risk of leaking the token in memory too, but same with encrypted storage. 

I don't think we can solve this with server id, because we can't write it to the memento.

Reply:
currentUri being sync is only a problem if controllerRegistration.add is called before it's resolved. 

However if it's resolved and the onDidChangeUri event were to fire, it doesn't matter. It's the same as if it was originally set to remote.

Oh but that still means isLocalLaunch wouldn't match. Hmm. 

So then we'd have to force async.

That ripples through the controllerRegistration add, onDidChangeUri, and on didChangeKernelFilter

- add being async seems possible.
- onDidChangeUri would should work if it actually gets passed the URI involved. 
- onDidChangeKerneFilter could be made async, doesn't really matter

Okay I think I can get rid of the synchronous of the currentUri.

Reply:
> I don't think we can solve this with server id, because we can't write it to the memento.

Given that we can store it in memento, i think we can solve it

Reply:
> Oh but that still means isLocalLaunch wouldn't match. Hmm.

My suggestion is:
* Create a property to keep track of isloca or remote (initially undefiend)
* When loading the first time, this property is empty and we read the value from memento to get the server id or the like (then we know whether its local or not)
* Subsequently when we update memento, we can update this property as well
* Then next time we can check if this property is not undefined & use the value in that property.
Thats how the serverConnectionType class worked, i.e. it solved this problem by having synchronous access via the `_isLocalLaunch` private property.

Reply:
hashjs is not cryptographically secure. We'd have to use something else. The security review team specifies this when we said we don't use it to store anything but just to hash stuff that isn't a password. We start using it to store passwords (which is what the token in the url is) we need to use something other than hashjs.